src/cpp/src/block_manager.hpp

```diff
@@ -134,131 +162,227 @@ class BlockAllocator {
     }
 
     size_t num_free_blocks() const {
-        return m_free_blocks.size() + m_evictor.num_blocks();
+        OPENVINO_ASSERT(m_num_layers == 0, "this overload may only be used when num_layers == 0");
+        return num_free_blocks(0);
+    }
+
+    size_t num_free_blocks(size_t layer_idx) const {
+        return m_free_blocks[layer_idx].size() + m_evictor.num_blocks();
     }
 
     bool can_allocate_blocks(size_t num_blocks) const {
-        return num_blocks <= num_free_blocks();
+        OPENVINO_ASSERT(m_num_layers == 0, "this overload may only be used when num_layers == 0");
```

Comment:
why do we need such limitation? 
Am I right that in case of KV cache eviction, all layers will evict the same number of blocks? If yes, in this case number of free blocks in allocator is the same for each layers.

Reply:
Refactored